### PR TITLE
[@babel/traverse]: Added types to argument of Scope#push.

### PR DESCRIPTION
**Pull request by cameron-martin**
Copied from the flow types (https://github.com/babel/babel/blob/61f2aed5b0d3ef869dca2f8e2673e46079df7d09/packages/babel-traverse/src/scope/index.js#L778) with one exception - kind was made optional, since this is assigned a default value (https://github.com/babel/babel/blob/61f2aed5b0d3ef869dca2f8e2673e46079df7d09/packages/babel-traverse/src/scope/index.js#L801) and is most likely a mistake in the flow types.

Existing tests give this change a bit of coverage - let me know if you'd like me to add more.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: URLs above.
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/babel__traverse/index.d.ts [hunk 1]
```diff
@@ -90,7 +90,13 @@ export class Scope {
 
     removeData(key: string): void;
 
-    push(opts: any): void;
+    push(opts: {
+        id: object,
+        init?: object,
+        unique?: boolean,
+        _blockHoist?: number,
+        kind?: "var" | "let",
```
Comment (rpetrich) [inline types/babel__traverse/index.d.ts]: I think const is also valid in this context?

File: types/babel__traverse/index.d.ts [hunk 2]
```diff
@@ -90,7 +90,13 @@ export class Scope {
 
     removeData(key: string): void;
 
-    push(opts: any): void;
+    push(opts: {
+        id: object,
+        init?: object,
+        unique?: boolean,
+        _blockHoist?: number,
```
Comment (rpetrich) [inline types/babel__traverse/index.d.ts]: Is this a supported API or for internal Babel use only?

File: types/babel__traverse/index.d.ts [hunk 3]
```diff
@@ -90,7 +90,13 @@ export class Scope {
 
     removeData(key: string): void;
 
-    push(opts: any): void;
+    push(opts: {
+        id: object,
+        init?: object,
```
Comment (rpetrich) [inline types/babel__traverse/index.d.ts]: Should be of type Expression?

File: types/babel__traverse/index.d.ts [hunk 4]
```diff
@@ -90,7 +90,13 @@ export class Scope {
 
     removeData(key: string): void;
 
-    push(opts: any): void;
+    push(opts: {
+        id: object,
```
Comment (rpetrich) [inline types/babel__traverse/index.d.ts]: Should be of type string?

**Review by rpetrich — COMMENTED** [review]
This is definitely better than the current typings, but I think these could be more precise and made some suggestions.

File: types/babel__traverse/index.d.ts (hunk 4 above)
Comment (cameron-martin) [reply types/babel__traverse/index.d.ts]: `id` and `init` get passed into the builder for `VariableDeclarator`, so presumably they take the same types as the builder - `LVal` and `Expression` respectively.

File: types/babel__traverse/index.d.ts (hunk 3 above)
Comment (cameron-martin) [reply types/babel__traverse/index.d.ts]: Correct, see above.

File: types/babel__traverse/index.d.ts (hunk 1 above)
Comment (cameron-martin) [reply types/babel__traverse/index.d.ts]: Yup, const works too.

File: types/babel__traverse/index.d.ts (hunk 2 above)
Comment (cameron-martin) [reply types/babel__traverse/index.d.ts]: Presumably internal - I can't think of another reason why an underscore would be used here.

**Comment by cameron-martin** [comment]
I've implemented those changes, and made another small change to use [`$ExpectError`](https://github.com/Microsoft/dtslint#write-tests) rather than `// The following should fail when uncommented`

**Review by marvinhagemeister — APPROVED** [review]
That's awesome! Thanks for making a PR and applying the suggested changes 👍💯

**Review by rpetrich — APPROVED** [review]

**Pull request merged**